assertj-core/src/main/java/org/assertj/core/api/DefaultAssertionErrorCollector.java

```diff
@@ -28,12 +30,13 @@ public class DefaultAssertionErrorCollector implements AssertionErrorCollector {
   private volatile boolean wasSuccess = true;
   private List<AssertionError> collectedAssertionErrors = synchronizedList(new ArrayList<>());
 
-  private AfterAssertionErrorCollected callback = this;
+  private Set<AfterAssertionErrorCollected> callbacks = new HashSet<>();
```

Comment:
We probably want an ordered collection here in case the ordering of callback is important, so a list or a linked hash set.
Why would we need a set here ? in case you want the registration process to be idempotent ? any use case in min ?

Reply:
Maybe I was too focused on my own case while choosing Collection. I used AOP and it seems like my `pointcut` catched creating the same soft assertion object twice. I dont remember details to be honest.
I expected this question and was ready for it, but I found a problem. According to JLS-15.27.4:

> the identity of the result of evaluating a lambda expression is unpredictable, and therefore identity-sensitive operations may produce different results in different implementations of the Java programming language, or even upon different lambda expression evaluations in the same implementation.

So it looks risky and as you said I'd suggest good old `ArrayList`. But should I make it as usual or like it done for `collectedErrors` field?
`private List<AssertionError> collectedAssertionErrors = synchronizedList(new ArrayList<>());`
Unfortunatelly I'm not good at concurrency issues.

Reply:
Safer to use a `synchronizedList`, won't be a performance issue since the callback registration won't happen often